### PR TITLE
Coverage instrumented binary

### PR DESCRIPTION
**Pull request by dlespiau**
The go toolchain doesn't seem to support building coverage instrumented binaries. We can use a [succession of tricks](http://damien.lespiau.name/2017/05/building-and-using-coverage.html) to make it happen.

Fortunately with this, we can very easily meet our 50% coverage target... and aim at 80-90% :)

### CODE REVIEW

File: .ci/go-no-os-exit.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+for f in $candidates; do
+	filename=`basename $f`
+	# skip exit.go where, the only file we should call os.Exit() from.
+	[[ $filename == "exit.go" ]] && continue
+	# skip main_test.go
+	[[ $filename == "main_test.go" ]] && continue
+	files="$f $files"
+done
+
+if grep -n os.Exit $files; then
```
Comment (jodh-intel) [inline .ci/go-no-os-exit.sh]: `if egrep -q '\<os.Exit\>' $files; then` ?

File: exit.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+var atexitFuncs []func()
+
+// atexit registers a function f that will be run when exit is called.
+func atexit(f func()) {
+	atexitFuncs = append(atexitFuncs, f)
+}
+
+// exit calls all atexit handlers before exiting the process with status.
+func exit(status int) {
```
Comment (jodh-intel) [inline exit.go]: Neat. I wonder if there is any value in printing out the list of registered functions using `reflect` here?

File: exit.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	for _, f := range atexitFuncs {
```
Comment (jodh-intel) [inline exit.go]: If this is really supposed to be like `atexit(3)`, you need to reverse the order so the first function registered gets called last.

File: .ci/go-no-os-exit.sh (hunk 1 above)
Comment (dlespiau) [inline .ci/go-no-os-exit.sh]: Why -q? we don't want it to be quiet but tell us where offending os.Exit is (file/line)

Comment (dlespiau) [inline .ci/go-no-os-exit.sh]: I guess \. as well

Comment (jodh-intel) [inline .ci/go-no-os-exit.sh]: yep, we need `-n`, not `-q` and should escape the dot as you say.

File: exit.go (hunk 3 above)
Comment (dlespiau) [inline exit.go]: That's a good idea, fixing!

**Comment by dlespiau** [comment]
PR updated:

- Rebased onto master
- exit() calls AtExit() handlers in reverse order
- Be more scripts in the os.Exit check for the CI test

**Comment by jodh-intel** [comment]
lgtm

**Comment by sboeuf** [comment]
lgtm

**Pull request merged**